### ISSUE TITLE
How to infer top-k output sequences?

### ISSUE THREAD

**Issue by MInner** [google/seq2seq#158]
Is there a way to infer let's say top 5 output sequences for each input sequence with corresponding log-likelihoods? I wish it was possible by looking into beam search outputs, but as far as I could tell from visualization, beam search stops whenever it finds first high probability sequence end symbol, so other high probability sequences in beam search output are "not finished", right?

**Comment by MInner** [comment]
Oops, I interpreted output incorrectly. Yes, one can extract k most likely outputs, I will put code here when it is ready.

**Comment by shubhamagarwal92** [comment]
@MInner are you thresholding the number of returned top k output predictions based on any probability criteria? I tried running your snippet (working perfectly fine) but I am getting different number of top K predictions for each of the sequences. Could you discuss how can we avoid making graphs so that it might run faster.

**Comment by MInner** [comment]
@shubhamagarwal92 , yes, I also noticed this behavior. My understanding is that it should extract all paths that match the criteria, and the code that performs actual beam search sometimes produces more then five paths that match this criteria.

The graph extraction code above is written not by me, but by the owner of this repo.

I have occasionally posted a slow version, check out the link above. It also pads the result with zero probability sequences and crops if there were too many of them.

**Comment by MInner** [comment]
Btw, here is a small thing; using a `zip()` [here](https://gist.github.com/MInner/5204e649f6a7b0541b232f1f0f9fc8ba#file-top_k_seq2seq-py-L108) considerably improves performance of this piece of code (and of original [generate_beam_viz.py](https://github.com/google/seq2seq/blob/master/bin/tools/generate_beam_viz.py#L110) too) because, as far as I understand, `beam_data` is a 1x1 numpy array with a dict inside, so getting an array from a map from a 1x1 numpy array and indexing it on every loop iteration is quite slow. @dennybritz , I wonder if it would be easier if I do a pull request or it is easier for you to change, because it is literally 4 lines of code?

**Comment by shubhamagarwal92** [comment]
@MInner I noticed you made changes to the original gist you posted. The new code is for a (hard coded) beam width of 5? Could you elaborate more about the changes, would the results have exactly top-k(5) predictions now?

I think that we can directly use the 'gather_tree' function defined [here](https://github.com/google/seq2seq/blob/b9aa3a628ddc4a99879722fb36522c1e7097e1a4/seq2seq/inference/beam_search.py#L88) as used for testing in [beam_search_test](https://github.com/google/seq2seq/blob/ffb2747664dad58b46b8088f93df9d7964a3fc54/seq2seq/test/beam_search_test.py) instead of making a graph. 

As I expected, I am getting fixed number of predictions (some incomplete due to early stopping). Would share the gist on request. 

What really boggles me is that using your previous code (based on [generate_beam_viz](https://github.com/google/seq2seq/blob/2500c26add91b079ca00cf1f091db5a99ddab9ae/bin/tools/generate_beam_viz.py)), I was able to get more complete sequences than the pre-defined beam width (ie if beam width was defined to be 20, I was getting around 20-25 top complete sequences for some of the source lines).  

@dennybritz could you comment on this.

**Comment by Scitator** [comment]
I rewrite the script, now it more like original one. And should work well with linux pipes.
https://gist.github.com/Scitator/f3cce81338705dd7fddce1cbeefa2311

**Comment by VatsalThrow** [comment]
How to run the above script after training to get the top 5 as output?
I tried running the script but see no output.

**Comment by shubhamagarwal92** [comment]
Hi Vatsal, 

Based on this script by @Scitator I have created this [gist](https://gist.github.com/shubhamagarwal92/731f471d5a88a8bdf4d3313b9cfb570b) which writes top k predictions as well as shell script which you have to configure for your use case. 

This code aslo creates a copy of source where each line is repeated as many times as number of beam predictions as well as maintaining index for each source line.

**Comment by VatsalThrow** [comment]
Thanks, this is helpfull.
